### PR TITLE
Implement GraphARM model from [Kong et al. 2023]

### PR DESCRIPTION
**Pull request by caio-freitas**
Implement Autoregressive Diffusion Model for Graph Generation model from Kong et al.

### CODE REVIEW

File: benchmarks/GraphARM/models.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+class DiffusionOrderingNetwork(nn.Module):
+    '''
+    at each diffusion step t, we sample from this network to select a node 
+    v_sigma(t) to be absorbed and obtain the corresponding masked graph Gt
+    '''
+    def __init__(self,
+                 node_feature_dim,
+                 num_node_types,
+                 num_edge_types,
```
Comment (caio-freitas) [inline benchmarks/GraphARM/models.py:16]: Add edge types for masked and empty connections

File: benchmarks/GraphARM/utils.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+
+# def node_decay_ordering(datapoint):
+#     # create random list of nodes
+#     return torch.randperm(datapoint.x.shape[0]).tolist()
+
+class NodeMasking:
+    def __init__(self, dataset):
+        self.dataset = dataset
+        self.NODE_MASK = dataset.x.unique().shape[0]
```
Comment (caio-freitas) [inline benchmarks/GraphARM/utils.py:9]: A way to mask nodes without needing the whole dataset would be great (-1 doesn't work with the embedding layer)

File: imitation/dataset/robomimic_graph_dataset.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        # result must be of shape (num_nodes, num_node_feats)
+        return node_feats
+    
+
+
+    def _get_edge_index(self):
+        # Adjacency matrix must be converted to COO format
+        # for now, all nodes connected to next node
+        edge_index = []
+        for idx in range(7): # TODO connectivity of objects to robot
```
Comment (caio-freitas) [inline imitation/dataset/robomimic_graph_dataset.py:94]: Add connectivity of objects to robot

File: imitation/dataset/robomimic_graph_dataset.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+        super().__init__(root=self._processed_dir, transform=None, pre_transform=None, pre_filter=None, log=True)
+
+    def _calculate_joints_positions(self, joints):
+        q = torch.tensor([joints]).to("cpu")
+        q.requires_grad_(True)
+        data = self.robot_fk.compute_forward_kinematics_all_links(q)
+        data = data[0]
+        joint_positions = []
+        # TODO add joint_quaternions
```
Comment (caio-freitas) [inline imitation/dataset/robomimic_graph_dataset.py:50]: Add orientations to node features

File: imitation/dataset/robomimic_graph_dataset.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                 action_keys,
+                 object_state_sizes,
+                 object_state_keys,
+                 pred_horizon=1,
+                 obs_horizon=1,
+                 action_horizon=1):
+        self.obs_keys = obs_keys
+        self.action_keys = action_keys
+        self.pred_horizon = pred_horizon        #|
+        self.obs_horizon = obs_horizon          #|} FIXED TO 1 in the graph dataset
```
Comment (caio-freitas) [inline imitation/dataset/robomimic_graph_dataset.py:27]: How can we expand to obs_horizon > 1? Maybe adding another edge type (temporal connection), so whole trajectories can be generated.

File: benchmarks/GraphARM/grapharm.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+            diffusion_trajectories.append(diffusion_trajectory)
+        return diffusion_trajectories
+
+    def train_step(
+            self,
+            train_data,
+            val_data,
+            M = 4, # number of diffusion trajectories to be created for each graph
+        ):
+
```
[the remark this answers is not in the document]
Comment (caio-freitas) [reply benchmarks/GraphARM/grapharm.py:80]: According to  [this Stackoverflow thread](https://stackoverflow.com/questions/48001598/why-do-we-need-to-call-zero-grad-in-pytorch)/[pytorch docs](https://pytorch.org/docs/stable/_modules/torch/autograd.html) the gradients are accumulated, which is "is convenient while training RNNs or when we want to compute the gradient of the loss summed over multiple mini-batches". 

Looks like we don't need to use zero_grad, and can calculate `optimizer.backwards()` on each term of the `reward`/`loss`.

File: benchmarks/GraphARM/grapharm.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+        with tqdm(train_data) as pbar:
+            for graph in pbar:
+                # preprocess graph
+                graph = self.preprocess(graph)
+                diffusion_trajectories = self.generate_diffusion_trajectories(graph, M)  
+                # predictions & loss
+                for diffusion_trajectory in diffusion_trajectories:
+                    G_0 = diffusion_trajectory[0]
+                    # optimizer.zero_grad()
+                    node_order = self.node_decay_ordering(G_0) # TODO check. This isn't available when sampling
```
Comment (caio-freitas) [inline benchmarks/GraphARM/grapharm.py:103]: Node order when sampling should come from uniform distribution. Due to permutation invariance, the generation ordering network is not needed. 
> the denoising network itself can be a proxy of the generation ordering

File: benchmarks/GraphARM/grapharm.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+                        # predict node type
+                        node_type_probs, edge_type_probs = self.denoising_network(G_pred)
+                        
+                        # sample node type
+                        node_type = torch.distributions.Categorical(probs=node_type_probs[node]).sample()
+
+                        # calculate loss
+                        p_O_v =  node_type_probs[node][node_type] + EPSILON # TODO add edges
+                        w_k = self.diffusion_ordering_network(G_tplus1)[node]
+                        loss = loss + (n_i/(len(diffusion_trajectory)-1))*torch.log(p_O_v)*w_k/M # cumulative, to join (k) from all previously denoised nodes
```
Comment (caio-freitas) [inline benchmarks/GraphARM/grapharm.py:128]: @Alonso94 @anindex I'd like your opinion on this loss function. In the paper they describe is as being a sum over batch elements (graphs, denoted by $i$), trajectories $m$, timesteps $t$ and $k \in \sigma_{\leq t}$, where $\sigma$ are the node orderings.

image

Do you think it makes sense to accumulate it twice as I'm doing here? (I'm not zero-ing the gradients instead of creating the sum)

**Comment by caio-freitas** [comment]
Stale since branch was renamed
